refactor(ApplicationRef): invoke APP_INITIALIZER after ApplicationRef before bootstrap

PatrickJS opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [x] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)
APP_INITIALIZER are invoked during creation of platform/app injectors

**What is the new behavior?**
invoke APP_INITIALIZER after creation of platform/app injectors

**Does this PR introduce a breaking change?**
- [x] Yes
- [ ] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...
invoke APP_INITIALIZER after ApplicationRef but before bootstrap

**Other information**:
possible close #9101

gdi2290 commented on modules/@angular/core/src/application_ref.ts:
```diff
@@ -324,7 +301,30 @@ export class ApplicationRef_ extends ApplicationRef {
+      var asyncInitDonePromise: any /** TODO #9100 */;
+      if (isPresent(inits)) {
+        for (var i = 0; i < inits.length; i++) {
+          var initResult = inits[i]();
```
this assumes the developer would use `useValue` rather than `useFactory` which is why the `initResult` is being invoked. It should determine if a function or promise is invoked otherwise devs who use `useFactory` must return a function

laco0416 replied:
why isn't this `init[i](injector)` ?

gdi2290 replied:
if you're using `useFactory` then you have to create it like this
```
{
  useFactory: (injectorFromDi) => {
    return (injectorPassedDown) => Promise.resolve();
  }
  deps: [Injector]
}
```

APP_INITIALIZER was only meant to be used with `useValue` where you return a function that returns a promise

laco0416 replied:
This makes sense totally. Thanks!

gdi2290 commented on modules/@angular/core/src/application_ref.ts:
```diff
@@ -324,7 +301,30 @@ export class ApplicationRef_ extends ApplicationRef {
 
-  waitForAsyncInitializers(): Promise<any> { return this._asyncInitDonePromise; }
+  waitForAsyncInitializers(injector): Promise<any> {
+    return this.run(() => {
```
I'm not sure if we need `.this.run` anymore since it's being invoked in `coreLoadAndBootstrap`'s `appRef.run(() => {}`

gdi2290 commented:
the travis errors are due to `MockApplicationRef#waitForAsyncInitializers` and checkPublicApi

gdi2290 commented:
the other test fails because there's a separation between when APP_INITIALIZER is invoked and when it's created in di so the test just needs to be refactored

vicb commented:
@gdi2290 @vsavkin what is the status of this PR ?

tbosch commented on modules/@angular/core/src/application_ref.ts:
```diff
@@ -324,7 +301,30 @@ export class ApplicationRef_ extends ApplicationRef {
   }
 
-  waitForAsyncInitializers(): Promise<any> { return this._asyncInitDonePromise; }
+  waitForAsyncInitializers(injector: Injector): Promise<any> {
```
This is not called for synchronous bootstrap, i.e. from `coreBootstrap`, which is used by offline compiled applications.

tbosch replied:
I.e. `coreBootstrap` needs to be changed to call this as well.
E.g. change the name to `runInitializers` and return

- a Promise if there were async initializers
- null if there were no async initializers

Then call this from `coreLoadAndBootstrap` as well as from `coreBootstrap`.

tbosch commented on modules/@angular/core/src/application_ref.ts:
```diff
@@ -196,7 +196,7 @@ export abstract class ApplicationRef {
    * are done.
    */
-  abstract waitForAsyncInitializers(): Promise<any>;
+  abstract waitForAsyncInitializers(injector: Injector): Promise<any>;
```
Why do we need to pass the `Injector` now? `ApplicationRef` already knows about it (via `ApplicationRef.injector`)

gdi2290 replied:
oops, yeah I can fix that

gdi2290 commented:
@vicb I just now updated it per changes requested by @tbosch

gdi2290 commented:
all green, even for web workers

gdi2290 commented on modules/@angular/core/src/application_ref.ts:
```diff
@@ -374,7 +356,34 @@ export class ApplicationRef_ extends ApplicationRef {
+          let init = inits[i];
+          let initResult = init;
+          if (isFunction(init)) {
+            initResult = init();
```
this allows for `useValue` and `useFactory`
```es6
{provide: APP_INITIALIZER, useValue: () => Promise.resolve('wat')}
```
```typescript
{
  provide: APP_INITIALIZER,
  useFactory: (appRef) => {
    if (true) {
      return Promise.resolve('wat') // you can also return () => Promise.resolve('wat') just as useValue
    }
    return  null;
  },
  deps: [ApplicationRef]
}
```

gdi2290 commented on modules/@angular/core/testing/mock_application_ref.ts:
```diff
@@ -26,7 +26,7 @@ export class MockApplicationRef extends ApplicationRef {
   run(callback: Function): any { return null; }
 
-  waitForAsyncInitializers(): Promise<any> { return null; }
+  runInitializers(): Promise<any>|any { return null; }
```
`null` type only works in typescript 2.0

gdi2290 commented on modules/@angular/platform-browser/test/browser/bootstrap_spec.ts:
```diff
@@ -242,14 +242,17 @@ export function main() {
+         expect(log.result()).toEqual('');
+
+         let appRef = a.get(ApplicationRef);
+         appRef.runInitializers();
```
we can't, probably shouldn't, rely on creating `ApplicationRef` to fire off the initializers anymore

gdi2290 commented on tools/public_api_guard/core/index.d.ts:
```diff
@@ -376,7 +376,7 @@ export interface ContentChildrenMetadataFactory {
 }
 
-export declare function coreBootstrap<C>(componentFactory: ComponentFactory<C>, injector: Injector): ComponentRef<C>;
+export declare function coreBootstrap<C>(componentFactory: ComponentFactory<C>, injector: Injector): ComponentRef<C> | Promise<ComponentRef<C>>;
```
we now have to wrap this in a `Promise.resolve(coreBootstrap(componentFactory))` to ensure it works if we don't know what APP_INITIALIZER are being used

tbosch replied:
No, this should always be synchronous. It should throw if there are async initializers.

gdi2290 replied:
ok I can update that

PatrickJS committed:
refactor(ApplicationRef): invoke APP_INITIALIZER after ApplicationRef_ before bootstrap

BREAKING CHANGE:
style(ApplicationRef_): parameter type for waitForAsyncInitializers

style(ApplicationRef): parameter type for waitForAsyncInitializers

tbosch commented:
Closing this as `APP_INITIALIZERS` will soon get deprecated in favor of the constructor of app modules. I.e. by using an app module, the constructor of the app module is created eagerly and can inject the `ApplicationRef`.
